Framework/Beamline/src/ComponentInfo.cpp

```diff
@@ -359,5 +507,180 @@ bool ComponentInfo::isStructuredBank(const size_t componentIndex) const {
+  for (size_t timeIndex = 0; timeIndex < other.m_scanIntervals->size();
+       ++timeIndex) {
+    if (!toMerge[timeIndex])
+      continue;
+    auto &scanIntervals = m_scanIntervals.access();
+    auto &positions = m_positions.access();
+    auto &rotations = m_rotations.access();
+    scanIntervals.push_back((*other.m_scanIntervals)[timeIndex]);
+    const size_t indexStart = other.linearIndex({0, timeIndex});
+    size_t indexEnd = indexStart + nonDetectorSize();
+    positions.insert(positions.end(), other.m_positions->begin() + indexStart,
+                     other.m_positions->begin() + indexEnd);
+    rotations.insert(rotations.end(), other.m_rotations->begin() + indexStart,
+                     other.m_rotations->begin() + indexEnd);
+    ++m_scanCounts;
+  }
+
+  m_detectorInfo->merge(*other.m_detectorInfo);
+}
+
+std::vector<bool>
+ComponentInfo::buildMergeIndicesSync(const ComponentInfo &other) const {
+  checkSizes(other);
+  std::vector<bool> merge(other.m_scanIntervals->size(), true);
+
+  for (size_t t1 = 0; t1 < other.m_scanIntervals->size(); ++t1) {
+    for (size_t t2 = 0; t2 < m_scanIntervals->size(); ++t2) {
+      const auto &interval1 = (*other.m_scanIntervals)[t1];
+      const auto &interval2 = (*m_scanIntervals)[t2];
+      if (interval1 == interval2) {
+        for (size_t compIndex = 0; compIndex < nonDetectorSize(); ++compIndex) {
+          const size_t linearIndex1 = other.linearIndex({compIndex, t1});
+          const size_t linearIndex2 = linearIndex({compIndex, t2});
+          checkIdenticalIntervals(other, linearIndex1, linearIndex2);
+        }
+        merge[t1] = false;
+      } else if ((interval1.first < interval2.second) &&
+                 (interval1.second > interval2.first)) {
+        failMerge("sync scan intervals overlap but not identical");
+      }
+    }
+  }
+  return merge;
+}
+
+void ComponentInfo::checkSizes(const ComponentInfo &other) const {
+  if (size() != other.size())
+    failMerge("size mismatch");
+  if (!m_scanIntervals || !other.m_scanIntervals)
+    failMerge("scan intervals not defined");
+}
+
+void ComponentInfo::checkIdenticalIntervals(
+    const ComponentInfo &other, const size_t linearIndexOther,
+    const size_t linearIndexThis) const {
+  if ((*m_positions)[linearIndexThis] != (*other.m_positions)[linearIndexOther])
+    failMerge("matching scan interval but positions differ");
+  if ((*m_rotations)[linearIndexThis].coeffs() !=
+      (*other.m_rotations)[linearIndexOther].coeffs())
+    failMerge("matching scan interval but rotations differ");
+}
+
+/**
+ * As part of the public API, the ComponentInfo treats any component to be the
+ * same whether detector or otherwise. However, internally we often need to know
+ * what the non-detector size is as composed non-detector members are managed as
+ * part of this type.
+ * @return size of the component info in terms of non-detectors.
+ */
+size_t ComponentInfo::nonDetectorSize() const {
```

Comment:
I thought the comments here would have clarified things, but since we only store data in the ComponentInfo, for non detector components, we often need to translate between a component index (includes detectors), and an index that we need to use to fetch something at an address.

For example, imagine you had a beamline with 5 detectors, 1 source, 1 sample, and 1 root node. That's 8 components in total. Calling `ComponentInfo::position(0)` get the first detector, `ComponentInfo::position(7)` gets the root node. However, we only internally store a vector sized 3 for the non-detector components. 

`nonDetectorSize()` is a private function to allow us to do translations for some of the problems above. 

I'm happy to change the name, but I don't think your suggestion makes it any clearer.